File: leetcode/hard/stamping-the-sequence.cpp
// https://leetcode.com/problems/stamping-the-sequence
//
// Same as the Rust version; see that for details.
// Complexity: runtime O(t^2), space O(1) ignoring output.

class Solution {
public:
    vector<int> movesToStamp(const string stamp, string target) const
    {
        const auto isUndone = [](const auto it, const auto endIt) {
            return all_of(it, endIt, [](const char c) { return c == '?'; });
        };
        vector<int> res;

        while (!isUndone(cbegin(target), cend(target))) {
            const auto prevResSize = size(res);

            for (size_t i = 0; i < size(target) - size(stamp) + 1; ++i) {
                auto it = begin(target) + i, endIt = it + size(stamp);
                if (isUndone(it, endIt))
                    continue;

                size_t j = 0;
                for (; j < size(stamp)
                       && (*(it + j) == stamp[j] || *(it + j) == '?');
                     ++j);

                if (j == size(stamp)) {
                    fill(it, endIt, '?');
                    res.push_back(i);
                }
            }

            if (prevResSize == size(res))
                return {};
        }

        reverse(begin(res), end(res));
        return res;
    }
};

File: leetcode/hard/stamping-the-sequence.rs
// https://leetcode.com/problems/stamping-the-sequence
//
// Complexity: runtime O(t^2) [target = "aaaaaaab", stamp = "ab"; this can be mitigated by
//                             reversing iteration direction for "i" each pass],
//             space O(1) ignoring output [O(t) otherwise];
//             where t == target.len().

impl Solution {
    pub fn moves_to_stamp(stamp: String, target: String) -> Vec<i32> {
        let (stamp, mut target) = (stamp.as_bytes(), target.into_bytes());
        let mut res = Vec::new();

        while target.iter().any(|&c| c != b'?') {
            let prev_res_len = res.len();

            for i in 0..target.len() - stamp.len() + 1 {
                let sub = &mut target[i..i + stamp.len()];

                if sub.iter().any(|&c| c != b'?')
                    && sub
                        .iter()
                        .zip(stamp.iter())
                        .all(|(&wc, &sc)| wc == sc || wc == b'?')
                {
                    sub.iter_mut().for_each(|c| *c = b'?');
                    res.push(i as _);
                }
            }
            if res.len() == prev_res_len {
                return Vec::new();
            }
        }

        res.reverse();
        res
    }
}
